src/replication.c

```diff
@@ -126,37 +126,215 @@ static int handle_mgmt_event_fd(replica_t *replica);
+		    "for replica(%p)\n", replica);
+		return (ret = -1);
+	}
+	REPLICA_LOG("Start_rebuild opcode sent for Replica ip:%s port:%d "
+	    "state:%d\n", replica->ip, replica->port, replica->state);
+	return ret;
+}
+
+/* Rebuild can be of two type:-
+ * 1. Rebuilding a replica from healthy replica
+ * 2. Rebuilding a replica from all other replica(Mesh rebuild)
+ * API does the following tasks:-
+ * 1. find if there is healthy replica available for rebuilding
+ * 2. find downgraded replica for rebuild
+ * 3. Call prepare_for_rebuild to on healthy replica or on all
+ * other replicas to get IP address and port for replicas.
  */
 void
 update_volstate(spec_t *spec)
 {
+	int ret = 0;
 	uint64_t max;
-	replica_t *replica;
+	replica_t *replica = NULL;
+	replica_t *target_replica = NULL;
+	replica_t *healthy_replica = NULL;
+
+	if (spec->rebuild_in_progress == true) {
+		REPLICA_NOTICELOG("Rebuild is already in progress "
+		    "on replica:%s\n", spec->volname);
+		return;
+	}
+
+	if (((spec->healthy_rcount + spec->degraded_rcount >=
+	    spec->consistency_factor) && (spec->healthy_rcount >= 1)) ||
+	    (spec->healthy_rcount  + spec->degraded_rcount >=
+	    MAX(spec->replication_factor - spec->consistency_factor + 1,
+	    spec->consistency_factor))) { 
 
-	if(((spec->healthy_rcount + spec->degraded_rcount >= spec->consistency_factor) &&
-		(spec->healthy_rcount >= 1))||
-		(spec->healthy_rcount  + spec->degraded_rcount 
-			>= MAX(spec->replication_factor - spec->consistency_factor + 1, spec->consistency_factor))) {
-		if (spec->ready == false)
-		{
-			max = 0;
-			TAILQ_FOREACH(replica, &spec->rq, r_next)
-				max = (max < replica->initial_checkpointed_io_seq) ?
-				    replica->initial_checkpointed_io_seq : max;
+		max = 0;
+		TAILQ_FOREACH(replica, &spec->rq, r_next) {
+			/* Find healthy replica */
+			if ((healthy_replica == NULL) &&
+			    (replica->state == ZVOL_STATUS_HEALTHY)) {
+				healthy_replica = replica;
+			}
+
+			/*
+			 * If we know master replica and healthy
+			 * replica then trigger rebuild
+			 */
+			if ((spec->master_replica != NULL) &&
+			    (healthy_replica != NULL))
+				break;
+			/*
+			 * Should not account io_seq from healthy replica
+			 * when trying to find out degraded replica
+			 */
+			if (replica->state == ZVOL_STATUS_HEALTHY)
+				continue;
+			if (max < replica->initial_checkpointed_io_seq) {
+				max = replica->initial_checkpointed_io_seq;
+				target_replica = replica;
+			}
+		}
 
+		/*
+		 * Increase io_seq only when we do not have healthy replica
+		 */	
+		if (healthy_replica == NULL) {
 			max = (max == 0) ? 10 : max + (1<<20);
 			spec->io_seq = max;
 		}
 		spec->ready = true;
 	} else {
 		spec->ready = false;
 	}
+
+	REPLICA_LOG("Healthy count:%d degraded count:%d consistency factor:%d"
+	    " replication factor:%d\n", spec->healthy_rcount,
+	    spec->degraded_rcount, spec->consistency_factor,
+	    spec->replication_factor);
+
+	if (target_replica != NULL) {
```

Comment:
rebuilding process should start only when spec is 'ready' to take IOs (basically the condition to make spec as 'ready' should be met) or atleast one healthy replica exists.

Reply:
oh ok.. got it.. you are right sir.
as you said, assert would be needed for any further modifications between these two chunks of code.

Reply:
or for better clarity (which currently gives impression that target_replica should be non-null for rebuild), how about moving this under the case where spec's ready is made as true?